File: pilsprog/zlomky.cpp
#include <bits/stdc++.h>
#define endl '\n'
#define nl '\n'

using namespace std;
typedef long long ll;
typedef vector<long long> vll;
typedef vector<int> vint;
typedef vector<bool> vbool;
typedef pair<int, int> pint;

int gcd(int a, int b) {
  a = abs(a);
  b = abs(b);
  if (a < b) {
    swap(a, b);
  }
  while (b != 0) {
    int temp = a % b;
    a = b;
    b = temp;
  }
  return a;
}

void solve() {
  int v;
  cin >> v;
  cin.ignore();
  for (int _ = 0; _ < v; ++_) {
    string input;
    getline(cin, input);
    stringstream ss(input);
    string frac1, op, frac2;
    ss >> frac1 >> op >> frac2;

    int a_num, a_den, b_num, b_den;
    sscanf(frac1.c_str(), "%d/%d", &a_num, &a_den);
    sscanf(frac2.c_str(), "%d/%d", &b_num, &b_den);

    if (a_den < 0) {
      a_num = -a_num;
      a_den = -a_den;
    }
    if (b_den < 0) {
      b_num = -b_num;
      b_den = -b_den;
    }

    int lcm_den = a_den * b_den / gcd(a_den, b_den);
    int aa = a_num * (lcm_den / a_den);
    int bb = b_num * (lcm_den / b_den);

    pair<int, int> res;
    if (op == "+") {
      res = make_pair(aa + bb, lcm_den);
    } else {
      res = make_pair(aa - bb, lcm_den);
    }

    int g = gcd(res.first, res.second);
    res.first /= g;
    res.second /= g;

    if (res.second == 1) {
      cout << res.first << endl;
    } else {
      cout << res.first << "/" << res.second << endl;
    }
  }
}

signed main() {
  solve();
  return 0;
}

File: pilsprog/zlomky.py
def gcd(a, b):
    a = abs(a)
    b = abs(b)
    if a < b:
        a, b = b, a
    while (b != 0):
        a %= b;
        a,b=b,a
    return a;

def solve():
    v = int(input())
    for _ in range(v):
        i = input()
        i = i.split(" ")
        a = tuple([int(x) for x in i[0].split("/")])
        if a[1] < 0:
            a = (-a[0], -a[1])
        b = tuple([int(x) for x in i[2].split("/")])
        if b[1] < 0:
            b = (-b[0], -b[1])
        aa = a[0] * b[1] // gcd(a[1],b[1])
        bb = b[0] * a[1] // gcd(a[1],b[1])
        if i[1] == "+":
            res = (aa + bb, a[1]*b[1]//gcd(a[1],b[1]))
        else:
            res = (aa - bb, a[1]*b[1]//gcd(a[1],b[1]))
        ress = (res[0]//gcd(res[0],res[1]),res[1]//gcd(res[0],res[1]))
        if ress[1] == 1:
            print(ress[0])
        else:
            print(f"{ress[0]}/{ress[1]}")


if __name__ == "__main__":
    solve()
